src/librustc_mir/interpret/terminator.rs

```diff
@@ -373,6 +373,20 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                         trace!("Caller has too many args over");
                         return err!(FunctionArgCountMismatch);
                     }
+                    // Don't forget to check the return type!
+                    if let Some(caller_ret) = dest {
+                        let callee_ret = self.eval_place(&mir::Place::Local(mir::RETURN_PLACE))?;
+                        if !Self::check_argument_compat(caller_ret.layout, callee_ret.layout) {
+                            return err!(FunctionRetMismatch(
+                                caller_ret.layout.ty, callee_ret.layout.ty
+                            ));
+                        }
+                    } else {
+                        // FIXME: The caller thinks this function cannot return. How do
+                        // we verify that the callee agrees?
+                        // On the plus side, the the callee ever writes to its return place,
+                        // that will be detected as UB (because we set that to NULL above).
```

Comment:
@eddyb any thought on this?

Reply:
Does `mir.local_decls[RETURN_PLACE].ty.is_uninhabited()` work for uninhabited types which are only uninhabited due to a private field and thus don't expose that uninhabitedness on the type level?

Reply:
I think yes, is that how rustc decides to not give a return place? I'll try adding a check along these lines.

Reply:
Except it should be `layout.abi.is_uninhabited()`, right?

Reply:
Yeah that seems to work.